protocol/disableInitializers-in-constructor.md

```diff
@@ -0,0 +1,51 @@
+# Always using `_disableInitializers()` in constructors
+
+## Context
+
+Currently, most OP Stack proxied contracts prevent direct calls to their `initialize` functions on the implementation contract by calling it with default parameter values in the constructor. Additionally, these contracts do not expose a public `reinitialize` function to prevent reinitialization attempts.
+
+## Problem Statement
+
+So far this does the job but has a few drawbacks, namely:
+
+- It's a manual process (per contract) given that `initialize` function's input arguments are different for each contract. Also, a change in the `initialize` function's arguments will require updating the call to `initialize` manually.
+- If a public `reinitialize` function is added in the future, calling the `initialize` function in the constructor will not prevent reinitialization.
+- It can get verbose for initializers with many arguments. Here's an [example](https://github.com/ethereum-optimism/optimism/blob/36b41baea45f07e8644ff778a1ffd81fb0ad9e77/packages/contracts-bedrock/src/L1/SystemConfig.sol#L152).
+
+## Proposed Solution
+
+A proposed solution is to always use `_disableInitializers()` in the constructor of the implementation contract instead. This will, like the current approach, prevent anyone from calling the `initialize` function directly on the implementation contract but also solves the drawbacks mentioned above.
+
+- It has the benefit of being a one-time manual change for each contract regardless of how many arguments the `initialize` function expects or if it changes in the future.
+- If a public `reinitialize` function is used in the future, `_disableInitializers()`, already called in the constructor, prevents reinitialization attempts on the implementation contract.
```

Comment:
It's not clear to my why the current approach of calling `initialize()` (as long as it has the `initializer` modifier on it) is different from calling `_disableInitializers()`?

Reply:
The major benefit is that `_disableInitializers()` is less verbose and does not need modification if initialize parameters change. 

But then also, there's a scenario where if `reinitializer` modifier is used in the future for some reason, the call to `initialize` in the constructor does not stop the implementation contract from being reinitialized after deployment. This is because `initialize` sets the initialized storage value to 1 which makes it reinitializable if the public function to do so exists but `_disableInitializers()` sets this value to `type(uint8).max`, preventing this entirely.

Reply:
> This is because initialize sets the initialized storage value to 1 which makes it reinitializable if the public function to do so exists but _disableInitializers() sets this value to type(uint8).max, preventing this entirely.
What do you mean by this, I don't get it. 
Does this means that `reinitializer()` should check that the slot value is `1` only and should revert if the slot is set to `type(uint8).max` ? 
I don't see it here:  https://github.com/OpenZeppelin/openzeppelin-contracts/blob/448efeea6640bbbc09373f03fbc9c88e280147ba/contracts/proxy/utils/Initializable.sol#L152

Reply:
It works if the slot is 1 but it can't be nested. For the max, it reverts if the version is > `type(uint64*).max` (which can be set manually via `reinitialize` or via `_disableInitializers()`